### PR TITLE
Allow to unset withdrawBlockers in undoCurrentSolution

### PR DESCRIPTION
**Pull request by josojo**
This PR enables a new solution submitter to undo all withdrawBlocks, which have been set by a previous solution.

### CODE REVIEW

**Review by nlordell — APPROVED** [review]

File: contracts/BatchExchange.sol [hunk 1]
```diff
@@ -540,11 +540,15 @@ contract BatchExchange is EpochTokenLocker {
...
                 revertRemainingOrder(owner, orderIndex, sellAmount);
-                subtractBalance(owner, tokenIdToAddressMap(order.buyToken), buyAmount);
+                address buyToken = tokenIdToAddressMap(order.buyToken);
+                subtractBalance(owner, buyToken, buyAmount);
+                undoWithdrawBlock(owner, buyToken);
                 emit TradeReversion(owner, orderIndex, sellAmount, buyAmount);
             }
             // subtract granted fees:
-            subtractBalance(latestSolution.solutionSubmitter, tokenIdToAddressMap(0), latestSolution.feeReward);
+            address solutionSubmitter = latestSolution.solutionSubmitter;
```
Comment (fleupold) [inline contracts/BatchExchange.sol:549]: For addition we have this logic exposed as it’s own method in the EpochTokenLocker (addBalanceAndDelayWithdraw). 

I think we should do the same for subtractions.

File: test/stablex/batch_exchange.js [hunk 2]
```diff
@@ -1515,6 +1515,106 @@ contract("BatchExchange", async accounts => {
       )
       assert.equal(0, (await batchExchange.currentPrices.call(2)).toString(), "CurrentPrice were not adjusted correctly")
     })
+    it("checks that withdraw blockers between solutions are reset", async () => {
```
Comment (fleupold) [inline test/stablex/batch_exchange.js:1518]: This is really testing an implementation detail. I think instead we should test that you can withdraw your funds even if the token was traded on your behalf in the first solution of the test (of it was undone in the second solution).

**Review by fleupold — APPROVED** [review]
I’m generally fine with this, but I’m not sure it’s worth it.
I think it would be worth mentioning the trade offs (extra gas) and what this achieves (attack that is possible in the current scenario). 

Also this is definitely something that needs to be flagged to the reviewers as adding last minute logic like this is a good candidate for bugs.

**Review by fleupold — CHANGES_REQUESTED** [review]
Let's review this again after we combine subtractBalance and undoWithdrawBlock.

**Comment by josojo** [comment]
We agreed that we will close this issue, as the attack for withdraw-locking is very very unlikely, and can always be prevented by canceling the orders.
Additionally, this new solution would increase the gas amount for submitting the non-first solution.

**Pull request closed without merging**